docs/developer/architecture/solutions.rst

```diff
@@ -0,0 +1,241 @@
+Solution deployment
+~~~~~~~~~~~~~~~~~~~
+
+As a solution admin, I want to be able to deploy a Solution through its
+UI, without any specific knowledge about Kubernetes.
+
+Solution upgrade/downgrade
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+As a solution admin, I want to be able to easily upgrade or downgrade a
+Solution using its UI without any specific knowledge about Kubernetes.
+
+Solution deletion
+~~~~~~~~~~~~~~~~~
+
+.. todo:: Open question:
+          do we want the user to be able to delete its solution ?
+          no ?
+          yes, but with some safety nest ?
+          Only with a tool from the CLI ?
+          If not, do we want to at least document the manual procedure ?
+
+Solution development
+~~~~~~~~~~~~~~~~~~~~
+
+As a developer, I want to have guidelines to follow to develop a Solution.
+
+Solution packaging
+~~~~~~~~~~~~~~~~~~
+
+As a developer, I want to have documentation to know how to package a Solution.
+
+Solution validation
+~~~~~~~~~~~~~~~~~~~
+
+As a developer, I want to be able to know that a packaged Solution
+follows the requirements and is valid using a CLI tool.
+
+Design Choices
+--------------
+
+Solution configuration
+~~~~~~~~~~~~~~~~~~~~~~
+
+We will use a ``SolutionsConfiguration`` object, stored in
+``/etc/metalk8s/solutions.yaml``, to declare which Solutions are available
+to the cluster, from the bootstrap node.
+
+Here is how it will look::
+
+    apiVersion: metalk8s.scality.com/v1alpha1
+    kind: SolutionsConfiguration
+    solutions:
+      - /solutions/storage_1.0.0.iso
+      - /solutions/storage_latest.iso
+      - /other_solutions/computing.iso
+
+There is no explicit information about what an archive contains.
+Instead, we want the archive itself to contain such information and to
+discover it at import time.
+
+This configuration file will be read by a Salt external pillar module,
+which will permit the consumption of it by Salt modules and states.
+
+Archive format
+~~~~~~~~~~~~~~
+
+Solution archives must use the `ISO-9660:1988`_ format, including `Rock Ridge`_
+and Joliet_ directory records. The character encoding must be UTF-8_. The
+conformance level is expected to be at most 3, meaning:
+
+- Directory identifiers may not exceed 31 characters (bytes) in length
+- File name + ``'.'`` + file name extension may not exceed 30 characters
+  (bytes) in length
+- Files are allowed to consist of multiple sections
+
+The generated archive should specify a volume ID, set to
+``{project_name} {version}``.
+
+.. _`ISO-9660:1988`: https://www.iso.org/obp/ui/#iso:std:iso:9660:ed-1:v1:en
+.. _`Rock Ridge`: https://en.wikipedia.org/wiki/Rock_Ridge
+.. _Joliet: https://en.wikipedia.org/wiki/Joliet_(file_system)
+.. _UTF-8: https://tools.ietf.org/html/rfc3629
+
+OCI Images registry
+~~~~~~~~~~~~~~~~~~~
+
+Every container images from Solution archive will be exposed as a single
+repository through MetalK8s registry. The name of this repository will be
+computed from the product information ``<NAME>-<VERSION>``.
```

Comment:
We may need to handle name collisions (in case `<VERSION>` is not unique...). To manage this, we will need to inject this information (where an OCI repository sits for a given version of a Solution) into the main components (UI, Operator) of each Solution instance.